### ISSUE TITLE
Custom client-side Content Type-based Form Controllers are not executed

### ISSUE THREAD

**Issue by russdanner** [craftercms/craftercms#5810]
CI: 5555

### Duplicates

- [X] I have searched the existing issues

### Latest version

- [X] The issue is in the latest released 4.0.x
- [ ] The issue is in the latest released 3.1.x

### Describe the issue

Custom client-side content type-based form controllers are not executed when configured for a given content type.

Example config (`/config/studio/config/content-types/component/schedule/config.xml`):
```
<content-type name="/component/schedule" is-wcm-type="true">
	<label>Schedule</label>
	<form>/component/schedule</form>
	<form-path>simple</form-path>
	<model-instance-path>NOT-USED-BY-SIMPLE-FORM-ENGINE</model-instance-path>
	<file-extension>xml</file-extension>
	<content-as-folder>false</content-as-folder>
	<previewable>false</previewable>
	<quickCreate>false</quickCreate>
	<quickCreatePath></quickCreatePath>
	<noThumbnail>true</noThumbnail>
	<image-thumbnail></image-thumbnail>
	<controller>true</controller>
	<paths>
		<includes>
			<pattern>^/site/components/schedules/.*</pattern>
		</includes>
	</paths>
</content-type>
```

Example Controller  (`/config/studio/config/content-types/component/schedule/form-controller.js`):
```
CStudioForms.FormControllers.ScheduleComponetController = CStudioForms.FormControllers.ScheduleComponetController ||  function() {
​
};
​
YAHOO.extend(CStudioForms.FormControllers.ScheduleComponetController, CStudioForms.FormController, {
​
    onBeforeSave : function () {
        var internalName = this.form.model['internal-name'];
        if(!internalName.endsWith("*")) {
            this.form.model['internal-name'] =  internalName + '*'
        }
        
        return true
    }
});
​
CStudioAuthoring.Module.moduleLoaded("/component/schedule-controller", CStudioForms.FormControllers.ScheduleComponetController );
```

There are two components of this issue:
- Saving a form definition for a content type removes the controller flag from config.xml
- The form engine does not attempt to load or run the form-controller.js associated to the content-type. 

### Steps to reproduce

N/A

### Relevant log output

```shell
N/A
```

### Screenshots and/or videos

N/A

**Comment by sumerjabri** [comment]
FE2

**Comment by sumerjabri** [comment]
This will be part of FE2, closing.

**Comment by russdanner** [comment]
Support for this was muted / removed due to the removal of the API it relies on
https://github.com/craftercms/studio-ui/blob/support/3.1.x/static-assets/components/cstudio-forms/forms-engine.js#L2788C1-L2823C11

**Comment by rart** [comment]
We need an api analog to `content-type/preview_image` to load the form controller.

Please create `/studio/api/2/configuration/content-type/form_controller?contentTypeId={typeId}&siteId={siteId}`

For example: `studio/api/2/configuration/content-type/preview_image?contentTypeId=%2Fcomponent%2Farticles-widget&siteId=editorial`

The form controller is a single JS file per content type with name `form-controller.js`, located at `{sandbox}/config/studio/content-types/{typeId}/form-controller.js`. The file is not mandatory, meaning it may not exist in certain content types. Must make sure the right content-type header is set so the browser will execute the file as if loading a regular JS file directly.

The API needs to be tight, specifically devoted to load this one file if present.

**Comment by sumerjabri** [comment]
@rart the API is ready for 4.x, see the OAS file for spec.

@jmendeza please send the same for v5, but include a unit test for v5 if possible.

**Comment by rart** [comment]
Note:
- We added the new controller file fetcher API so existing installations using the feature should remain unchanged.
- I'm attaching the `form-controller.js` I used for developing/testing. This could be used for documentation. I did not find this feature documented. For FE2, should we carry this forward, we must discuss the impact of existing capabilities of the feature (e.g. isFieldRelevant will cause a required field by definition to not be rendered and skipped).
- Apart from the feature being restored, there were two additional issues that at a glance, seem to come from even when the feature existed:
  - [Fixed] Saving the content type would wipe the `controller` property of the `config.xml` that informs the forms engine to load the controller in the first place.
  - [Fixed] Returning false on the `onBeforeSave`, would leave the form stuck since the buttons would remain disabled. Now, after returning `false`, it will still stop the save, but the button will be reenabled to allow retrying.
  - Stopping the save on the `onBeforeSave` callback of the controller doesn't show any kind of feedback. The controller would have the responsibility to issue some sort of dialog to inform the user of what is happening.

Q: Is this something we should include in the test harness?
Q: Is this something we should add docs for?

[form-controller.js.txt](https://github.com/user-attachments/files/18795356/form-controller.js.txt)

**Comment by sumerjabri** [comment]
@rart, schedule a meeting with Russ and me to discuss this.

**Comment by russdanner** [comment]
@rart I noticed that the structure of the code in the example controller you provided is a bit different from what I think is provided as the example in the ticket, and why users may have installed it. Will users need to modify (even slightly) existing controllers?

**Comment by rart** [comment]
@russdanner no. The API did not change. I just made things more inline with what I'd like to see in practice, but it's equivalent.

**Comment by russdanner** [comment]
Confirmed in 4.3
Note: Save as Draft does NOT reflect onSave changes made by the test script. 
This is equivalent to the behavior of the same script in previous versions.

**Comment by russdanner** [comment]
@rart  @jvega190 How is this handled/what are the documentation updates required for this with FE2?

**Comment by russdanner** [comment]
Usecases include:
- automatically setting a field value based on some external factor (date, weekday, api call etc)
- Controlling a drop down's values based on other drop downs (this is proibably better as custom control)
- Blocking the save
- Disabling access to a field based on a condition (like user role or other field value)
- Hiding fields for similar reasons
